### PR TITLE
Deansie/exercise4

### PR DESCRIPTION
**Pull request by Deansie**
## Pull Request Type
Please check the type of change your PR introduces:
- [x] Feature (a new feature for the project)  
- [ ] Fix (a bug fix)  
- [ ] Chore (no production code change)  
- [x] Refactor (refactoring production code)  
- [ ] Docs (documentation changes)  
- [ ] Style (formatting only)  
- [x] Test (adding/refactoring tests)  

## Description
This PR refactors the **Warehouse Product Manager** application to implement the Builder and Repository design patterns, as per issue #71. The application now uses a renamed `ProductService` class (formerly Warehouse), a Product record with a Builder for validation, and a `ProductRepository` for data access. 

### Key Features
- **Builder Pattern**: Implemented a Builder in the Product record for immutable object creation with validation.
- **Repository Pattern**: Created `ProductRepository` interface and `InMemoryProductRepository` implementation. Relocated product collection from ProductService to the repository. Added dependency injection via ProductService constructor.
- **ProductService**: Renamed from `Warehouse`, removed product collection and updated all the methods to use the repository.
- **CLI**: Update App to initialize ProductService with InMemoryProductRepository. 

## How Has This Been Tested?
- **JUnit 5 Tests**: Cover all `ProductService` methods with success, failure, and edge cases.  
- **Manual CLI Testing**: Verified adding, updating, querying, and error handling with both valid and invalid inputs.  
- **Production Mode**: Confirmed navigation, menu loop, and expected outputs when running via `org.example.App`.  

## Linked issues
This PR closes issue #71 .

### CODE REVIEW

File: src/main/java/org/example/App.java [hunk 1]
```diff
@@ -1,7 +1,180 @@
...
+            System.out.print("Enter Name: ");
+            String name = scanner.nextLine().trim();
+            System.out.print("Enter Category (ELECTRONICS, BOOKS, CLOTHING, FOOD, TOYS): ");
+            String categoryInput = scanner.nextLine().trim().toUpperCase();
+            Category category = Category.valueOf(categoryInput);
+            System.out.print("Enter Rating (0-10): ");
+            int rating = Integer.parseInt(scanner.nextLine().trim());
+            System.out.print("Enter Created Date (YYYY-MM-DD): ");
+            LocalDate createdDate = LocalDate.parse(scanner.nextLine().trim());
+            Product product = new Product.Builder()
```
Comment (kappsegla) [inline src/main/java/org/example/App.java:77]: An alternative could be a static method on Product returning a builder object to avoid  using new

File: src/main/java/org/example/service/ProductService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.time.LocalDate;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class ProductService {
+    private final ProductRepository productRepository;
+
+    public ProductService(ProductRepository productRepository) {
```
Comment (kappsegla) [inline src/main/java/org/example/service/ProductService.java:15]: Looks right with a dependency injection using constructor injection.

**Review by kappsegla — APPROVED** [review]
Looks to implement the requested functionality.   Checked the design patterns used and not the rest of the code since that has been reviewed before.
